### ISSUE TITLE
epoll_ctl assertion failure when Astrill VPN client installed

### ISSUE THREAD

**Issue by k-stachowiak** [zeromq/libzmq#4221]
# Issue description

I found that the basic example will crash if the Astrill VPN client is installed. It fails on an `epoll_ctl()` result assertion.
I don't need to be connected anywhere - just installing the VPN client seems to cause the problem. After I uninstall the VPN client the issue goes away.

On another note, when I put a breakpoint on the failing assertion, I noticed that it doesn't crash on the first two hits, which seem to happen on the main thread.
Only after the worker thread hits it, the crash occurs, which can be seen in the back trace.

I found the issue #4084 which seems to be similar, but the backtrace is different, and I think I have a clear path of reproduction here, hence I'm raising a new ticket. Also I upgraded my OS to the latest version, which was suggested in the other ticket, and the issue persists.

# Environment

* libzmq version (commit hash if unreleased): 4.3.4
* OS: Windows 10 (Version 10.0.18363 Build 18363)
* Compiler: Visual Studio 2019 (both: library and the example built manually)

# Minimal test code / Steps to reproduce the issue

    //  Hello World server
    #include <zmq.h>
    #include <string.h>
    #include <stdio.h>  
    #include <assert.h>
    
    #ifdef _WIN32
    #include <Windows.h>
    #else
    #include <unistd.h>
    #endif
    
    int main(void)
    {
        //  Socket to talk to clients
        void* context = zmq_ctx_new();
        void* responder = zmq_socket(context, ZMQ_REP);
        int rc = zmq_bind(responder, "tcp://*:5555");
        assert(rc == 0);
    
        while (1) {
            char buffer[10];
            zmq_recv(responder, buffer, 10, 0);
            printf("Received Hello\n");
            Sleep(1);          //  Do some 'work'
            zmq_send(responder, "World", 5, 0);
        }
        return 0;
    }

# What's the actual result? (include assertion message & call stack if applicable)

A crash occurs when we call `zmq_recv()`, and the call stack is:

    >    libzmq-v142-mt-gd-4_3_4.dll!zmq::zmq_abort(const char * errmsg_=0x00c237c0) Line 84    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::epoll_t::add_fd(unsigned int fd_=608, zmq::i_poll_events * events_=0x00bfadf8) Line 100    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::io_object_t::add_fd(unsigned int fd_=608) Line 66    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::stream_listener_base_t::process_plug() Line 70    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::object_t::process_command(const zmq::command_t & cmd_={...}) Line 87    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::io_thread_t::in_event() Line 92    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::epoll_t::loop() Line 206    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!zmq::worker_poller_base_t::worker_routine(void * arg_=0x00bdc740) Line 146    C++    Symbols loaded.
         libzmq-v142-mt-gd-4_3_4.dll!thread_routine(void * arg_=0x00bdc76c) Line 55    C++    Symbols loaded.
         ucrtbased.dll!thread_start<unsigned int (__stdcall*)(void *),1>(void * const parameter=0x00bdfe08) Line 97    C++    Non-user code. Symbols loaded.

The crash originates from `errno_assert()` failure in the function:

    zmq::epoll_t::handle_t zmq::epoll_t::add_fd (fd_t fd_, i_poll_events *events_)
    {
        check_thread ();
        poll_entry_t *pe = new (std::nothrow) poll_entry_t;
        alloc_assert (pe);
    
        //  The memset is not actually needed. It's here to prevent debugging
        //  tools to complain about using uninitialised memory.
        memset (pe, 0, sizeof (poll_entry_t));
    
        pe->fd = fd_;
        pe->ev.events = 0;
        pe->ev.data.ptr = pe;
        pe->events = events_;
    
        const int rc = epoll_ctl (_epoll_fd, EPOLL_CTL_ADD, fd_, &pe->ev);
        errno_assert (rc != -1); // <-- CRASHES HERE
    
        //  Increase the load metric of the thread.
        adjust_load (1);
    
        return pe;
    }

# What's the expected result?

The library not crashing.
A workaround of configuring the library differently could work as well (e.g. can we disable the `ZMQ_IOTHREAD_POLLER_USE_EPOLL` to work around it).

**Comment by k-stachowiak** [comment]
I did some more digging.

I have no experience with the ZMQ internals, but looking at it as a black box, I discovered the following.
The first commit, where I could observe crashes with the Astrill VPN client present was 66d0f3511fe8b2e6393606990a3d66a846672f69.
At that point the example program crashes at the `zmq_socket()` call. The commit message states it's about implementing IPC transport for Windows.

However, several commits later, at ca552d563ff358094485c9ade83159dc6922ff4f we have a change that "fixes" the issue, but the crash occurs later, at the `zmq_recv()` call. This commit is about fixing problems with TCP loopback connections by using domain sockets if available for listener sockets.

Our application is 32-bit, so, while at it, I checked both 32- and 64-bit builds and they both fail consistently.

**Comment by k-stachowiak** [comment]
I did some more digging.

I found a similar issue in another project, here: https://github.com/python-trio/trio/issues/1659

It looks like the VPN software may be blocking certain operations for security reasons. In the linked issue, they resolved it by changing the base socket acquisition algorithm to fall back to `SIO_BSP_HANDLE_POLL` operation, should the `SIO_BASE_HANDLE` operation fail.
BTW, the base handle acquisition only fails for the TCP socket, but not others.

I implemented a quick fallback mechanism like this:

    diff --git a/external/wepoll/wepoll.c b/external/wepoll/wepoll.c
    index 3d301410..49eb12c9 100644
    --- a/external/wepoll/wepoll.c
    +++ b/external/wepoll/wepoll.c
    @@ -2156,6 +2156,10 @@ tree_node_t* tree_root(const tree_t* tree) {
     #define SIO_BASE_HANDLE 0x48000022
     #endif
    
    +#ifndef SIO_BSP_HANDLE_POLL
    +#define SIO_BSP_HANDLE_POLL 0x4800001D
    +#endif
    +
     int ws_global_init(void) {
       int r;
       WSADATA wsa_data;
    @@ -2179,8 +2183,19 @@ SOCKET ws_get_base_socket(SOCKET socket) {
                    sizeof base_socket,
                    &bytes,
                    NULL,
    -               NULL) == SOCKET_ERROR)
    -    return_map_error(INVALID_SOCKET);
    +               NULL) != SOCKET_ERROR)
    +    return base_socket;
    +
    +  if (WSAIoctl(socket,
    +               SIO_BSP_HANDLE_POLL,
    +               NULL,
    +               0,
    +               &base_socket,
    +               sizeof base_socket,
    +               &bytes,
    +               NULL,
    +               NULL) != SOCKET_ERROR)
    +    return base_socket;
    
    -  return base_socket;
    +  return_map_error(INVALID_SOCKET);
     }

And the chrash is no longer occuring. I will start incorporating it into our application, but in the meantime, maybe somebody here could comment on this solution. It looks like we must assume the socket will only be used for polling and not selecting, but maybe that's a non-issue (I'm still far from begin an expert). Does it make sense?

**Comment by bluca** [comment]
Could you please report this on https://github.com/piscisaureus/wepoll ? I don't want to deviate from the upstream version, since it's an external implementation

**Comment by k-stachowiak** [comment]
Thanks for your reply. I went to their GitHub site to do just that, but decided to check the most recent version before I post an issue. Apparently they have already addressed the problem in a more structured way than mine, [see here](https://github.com/piscisaureus/wepoll/blob/dist/wepoll.c#L2236).

Is there a chance of these changes gating into your tree any time soon? I don't mean a release tag, as we may work off of any git ref, but it would greatly simplify incorporating the fix in our app :)

**Comment by bluca** [comment]
Can you test https://github.com/zeromq/libzmq/pull/4228 ? I don't think there's any coverage of wepoll builds in the CI

**Comment by k-stachowiak** [comment]
Thanks for the blazing fast reaction! I checked the new tip of `master` with my test application and it doesn't crash now. To be extra sure, I also checked out a ref from before the change and that still crashed, so I can confirm, that #4228 indeed fixes the issue in my case.
